app/main/controllers/clusters.py

```diff
@@ -0,0 +1,225 @@
+    def update_cluster_name(user_id, cluster_id):
+        cluster = get_authorized_cluster(user_id, cluster_id)
+        body = request.get_json()
+        if body is None:
+            abort(400)
+
+        new_name = body.get('name', None)
+        if new_name is None:
+            abort(400)
+
+        cluster.name = new_name
+        cluster.update()
+
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        })
+
+    '''
+    Creates a new cluster from scratch (two sources come near each other)
+    '''
+    @app.route('/clusters/create_new', methods=['POST'])
+    @requires_auth('create:clusters')
+    def create_cluster_from_scratch(user_id):
+        # Check to ensure user is authorized to make changes
+
+        body = request.get_json()
+        if body is None:
+            abort(400)
+
+        item1_id = body.get('item1_id', None)
+        item2_id = body.get('item2_id', None)
+        if item1_id is None or item2_id is None:
+            abort(400)
+
+        # Get sources and create them if necessary
+        item1 = get_authorized_item(user_id, item1_id)
+        item2 = get_authorized_item(user_id, item2_id)
+        if item1 is None or item2 is None:
+            abort(400)
+
+        # Clusters should start with same parent cluster
+        if item1.cluster != item2.cluster:
+            abort(400)
+
+        x = body.get('x_position', None)
+        y = body.get('y_position', None)
+        name = body.get('name', None)
+        if not name:
+            abort(400)
+
+        cluster = Cluster(name=name, x_position=x, y_position=y)
+        if item1.cluster is not None:
+            cluster.parent_cluster = item1.cluster
+        else:
+            cluster.project = item1.project
+        cluster.insert()
+        cluster.child_items.append(item1)
+        cluster.child_items.append(item2)
+        cluster.update()
+        status_code = 201
+
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        }), status_code
+
+    @app.route('/clusters/<int:cluster_id>/sources/<int:item_id>',
+               methods=['PATCH'])
+    @requires_auth('update:clusters')
+    def add_to_existing_cluster(user_id, cluster_id, item_id):
+        cluster = get_authorized_cluster(user_id, cluster_id)
+        item = get_authorized_item(user_id, item_id)
+
+        if item.cluster == cluster:
+            abort(400)
+        if cluster.parent_cluster is not None:
+            if item.cluster != cluster.parent_cluster:
+                abort(400)
+
+
+        item.cluster = cluster
+
+        item.update()
+        cluster.child_items.append(item)
+        cluster.update()
+        status_code = 201
+
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        }), status_code
+
+    @app.route('/clusters/<int:cluster_id>/sources/<int:item_id>/remove',
+               methods=['PATCH'])
```

Comment:
Same as the previous one, this endpoint doesn't indicate what it actually does.
```suggestion
    @app.route('/clusters/<int:cluster_id>/items/<int:item_id>',
               methods=['DELETE'])
```
You don't need the last `remove` since it's indicated by the DELETE action. It's clear that the item isn't deleted completely because you start with `/clusters`, which indicates that you're just removing from that cluster. If you wanted to delete the item, you would just call DELETE `/items/{item_id}`